File: VB/Core/NLog.Target.LogViewVB.Core/DataStoreLoggerTarget.vb
Imports LogViewerVB.Core
Imports Microsoft.Extensions.DependencyInjection
Imports Microsoft.Extensions.Logging
Imports Microsoft.Extensions.Options
Imports NLog.Targets
Imports MsLogLevel = Microsoft.Extensions.Logging.LogLevel

<Target("DataStoreLogger")>
Public Class DataStoreLoggerTarget : Inherits TargetWithLayout

#Region "Fields"

    Private _dataStore As ILogDataStore
    Private _config As DataStoreLoggerConfiguration

#End Region

#Region "methods"

    Protected Overrides Sub InitializeTarget()

        ' we need to inject dependencies
        Dim serviceProvider As IServiceProvider = ResolveService(Of IServiceProvider)()

        ' reference the shared instance
        _dataStore = serviceProvider.GetRequiredService(Of ILogDataStore)

        ' load the config options
        Dim options As IOptionsMonitor(Of DataStoreLoggerConfiguration) _
            = serviceProvider.GetService(Of IOptionsMonitor(Of DataStoreLoggerConfiguration))

        _config = If(options Is Nothing, New DataStoreLoggerConfiguration(), options.CurrentValue)

        MyBase.InitializeTarget()

    End Sub

    Protected Overrides Sub Write(logEvent As LogEventInfo)

        ' cast NLog Loglevel to Microsoft LogLevel type
        Dim logLevel As MsLogLevel = [Enum].ToObject(GetType(MsLogLevel), logEvent.Level.Ordinal)

        ' format the message
        Dim message As String = RenderLogEvent(Layout, logEvent)

        ' retrieve the EventId
        Dim eventId As EventId = logEvent.Properties("EventId")

        If eventId.Id = 0 AndAlso _config.EventId.Id <> 0 Then
            eventId = _config.EventId
        End If

        Dim exMessage As String = String.Empty

        If logEvent.Exception IsNot Nothing Then
            If String.IsNullOrEmpty(logEvent.Exception.Message) Then
                If logLevel = MsLogLevel.Error AndAlso message IsNot Nothing Then
                    exMessage = message.ToString()
                End If
            Else
                exMessage = logEvent.Exception.Message
            End If
        End If


        ' add log entry
        _dataStore.AddEntry(New LogModel() With
        {
            .Timestamp = Date.UtcNow,
           .LogLevel = logLevel,
           .EventId = eventId,
           .State = message,
           .Exception = exMessage,
           .Color = _config.Colors(logLevel)
        })

        Debug.WriteLine($"--- [{logLevel.ToString()(0.3)}] {message} - {If(String.IsNullOrWhiteSpace(exMessage), "no error", exMessage)}")

        MyBase.Write(logEvent)

    End Sub

#End Region

End Class

File: CSharp/Core/NLog.Target.LogView.Core/DataStoreLoggerTarget.cs
using LogViewer.Core;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using NLog.Targets;
using System.Diagnostics;
using MsLogLevel = Microsoft.Extensions.Logging.LogLevel;

namespace NLog.Target.LogView.Core;

[Target("DataStoreLogger")]
public class DataStoreLoggerTarget : TargetWithLayout
{
    #region Fields

    private ILogDataStore? _dataStore;
    private DataStoreLoggerConfiguration? _config;

    #endregion

    #region methods

    protected override void InitializeTarget()
    {
        // we need to inject dependencies
        IServiceProvider serviceProvider = ResolveService<IServiceProvider>();

        // reference the shared instance
        _dataStore = serviceProvider.GetRequiredService<ILogDataStore>();

        // load the config options
        IOptionsMonitor<DataStoreLoggerConfiguration>? options
            = serviceProvider.GetService<IOptionsMonitor<DataStoreLoggerConfiguration>>();

        _config = options?.CurrentValue ?? new DataStoreLoggerConfiguration();

        base.InitializeTarget();
    }

    protected override void Write(LogEventInfo logEvent)
    {
        // cast NLog Loglevel to Microsoft LogLevel type
        MsLogLevel logLevel = (MsLogLevel)Enum.ToObject(typeof(MsLogLevel), logEvent.Level.Ordinal);

        // format the message
        string message = RenderLogEvent(Layout, logEvent);

        // retrieve the EventId
        EventId eventId = (EventId)logEvent.Properties["EventId"];

        // add log entry
        _dataStore?.AddEntry(new()
        {
            Timestamp = DateTime.UtcNow,
            LogLevel = logLevel,
            // do we override the default EventId if it exists?
            EventId = eventId.Id == 0 && (_config?.EventId.Id ?? 0) != 0 ? _config!.EventId : eventId,
            State = message,
            Exception = logEvent.Exception?.Message ?? (logLevel == MsLogLevel.Error ? message : ""),
            Color = _config!.Colors[logLevel],
        });
        
        Debug.WriteLine($"--- [{logLevel.ToString()[..3]}] {message} - {logEvent.Exception?.Message ?? "no error"}");
    }

    #endregion
}